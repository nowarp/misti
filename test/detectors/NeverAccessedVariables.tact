fun test1(): Int {
    let a: Int = 20;
    if (true) { // Bad: write-only variable
        a = 42;
    }
    return 42;
}
// suppress SuspiciousLoop
//fun test2() {
//    while(true) { let a: Int = 0; }
//}

fun test3(): Int {
    let a: Int = 20; // read-only variable
    let b: Int = 22; // read-only variable
    let c: Int = 24;
    if (true) { // no `neverAccessed` false positive: all are used
        c = a + b;
    }
    return c;
}

fun test4(): Int {
    let a: Int = 20; // never-accessed variable
    let b: Int = 22; // read-only variable
    return b;
}

contract FieldTest1 {
    f1: Address;
    f2: Int as uint256 = 0;
    foreachmap: map<Address, Address>;

    init() {
        self.f1 = sender();
    }

    get fun use_f1(): Address {
        return self.f1;
    }

    fun forEachUse() {
        // No warning: foreachmap is used
        let foreachmap: map<Address, Address> = self.foreachmap;
        foreach (_, _ in foreachmap) {}
    }
}

contract FieldTest2 {
    f1: Address;
    init() {
        self.f1 = sender();
    }
}

contract ConstantFieldTest {
    const val: Int = ton("0.01");
    fun test() {
        send(SendParameters{
            to: sender(),
            bounce: true,
            value: self.val,
            mode: SendRemainingValue + SendIgnoreErrors
        });
    }
}


contract TestContract {
    receive() {
        let test: Int = 123; // should be reported
    }
}

contract TestContract2 {
    receive() {
        let _test: Int = 123; // suppressed: no warning
    }
}

