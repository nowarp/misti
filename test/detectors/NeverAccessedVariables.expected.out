[MEDIUM] NeverAccessedVariables: Field f2 is never used
test/detectors/NeverAccessedVariables.tact:31:5:
  30 |     f1: Address;
> 31 |     f2: Int as uint256 = 0;
           ^
  32 |     foreachmap: map<Address, Address>;
Help: Consider creating a constant instead of field
See: https://nowarp.io/tools/misti/docs/detectors/NeverAccessedVariables

[MEDIUM] NeverAccessedVariables: Write-only variable: a
test/detectors/NeverAccessedVariables.tact:2:5:
  1 | fun test1(): Int {
> 2 |     let a: Int = 20;
          ^
  3 |     if (true) { // Bad: write-only variable
Help: The variable value should be accessed
See: https://nowarp.io/tools/misti/docs/detectors/NeverAccessedVariables

[MEDIUM] NeverAccessedVariables: Variable a is never accessed
test/detectors/NeverAccessedVariables.tact:24:5:
  23 | fun test4(): Int {
> 24 |     let a: Int = 20; // never-accessed variable
           ^
  25 |     let b: Int = 22; // read-only variable
Help: Consider removing the variable
See: https://nowarp.io/tools/misti/docs/detectors/NeverAccessedVariables

[MEDIUM] NeverAccessedVariables: Variable test is never accessed
test/detectors/NeverAccessedVariables.tact:71:9:
  70 |     receive() {
> 71 |         let test: Int = 123; // should be reported
               ^
  72 |     }
Help: Consider removing the variable
See: https://nowarp.io/tools/misti/docs/detectors/NeverAccessedVariables